src/QsCompiler/Transformations/CallGraph/CallGraphWalker.cs

```diff
@@ -0,0 +1,507 @@
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+{
+    using ConcreteGraphBuilder = CallGraphBuilder<ConcreteCallGraphNode, ConcreteCallGraphEdge>;
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using Range = DataTypes.Range;
+    using SimpleGraphBuilder = CallGraphBuilder<SimpleCallGraphNode, SimpleCallGraphEdge>;
+    using TypeParameterResolutions = ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    /// <summary>
+    /// This transformation walks through the compilation without changing it, populating a given call graph as it does.
+    /// </summary>
+    internal static class BuildCallGraph
+    {
+        /// <summary>
+        /// Populates the given graph based on the given callables.
+        /// </summary>
+        public static void PopulateSimpleGraph(SimpleGraphBuilder graph, IEnumerable<QsCallable> callables) => SimpleCallGraphWalker.PopulateSimpleGraph(graph, callables);
+
+        /// <summary>
+        /// Populates the given graph based on the given compilation. This will produce a call graph that
+        /// contains all relationships amongst all callables in the compilation.
+        /// </summary>
+        public static void PopulateSimpleGraph(SimpleGraphBuilder graph, QsCompilation compilation) => SimpleCallGraphWalker.PopulateSimpleGraph(graph, compilation);
+
+        /// <summary>
+        /// Populates the given graph based on the given compilation. Only the compilation's entry points and
+        /// those callables that the entry points depend on will be included in the graph.
+        /// </summary>
+        public static void PopulateTrimmedGraph(SimpleGraphBuilder graph, QsCompilation compilation) => SimpleCallGraphWalker.PopulateTrimmedGraph(graph, compilation);
+
+        /// <summary>
+        /// Populates the given graph based on the given compilation. Only the compilation's entry points and
+        /// those callables that the entry points depend on will be included in the graph.
+        /// </summary>
+        public static void PopulateConcreteGraph(ConcreteGraphBuilder graph, QsCompilation compilation) => ConcreteCallGraphWalker.PopulateConcreteGraph(graph, compilation);
+
+        private static class BaseCallGraphWalker<TGraph, TNode, TEdge>
+            where TGraph : CallGraphBuilder<TNode, TEdge>
+            where TNode : CallGraphNodeBase
+            where TEdge : CallGraphEdgeBase
+        {
+            public abstract class TransformationState
+            {
+                internal TNode CurrentNode;
+                internal readonly TGraph Graph;
+
+                // The type parameter resolutions of the current expression.
+                internal IEnumerable<TypeParameterResolutions> ExpTypeParamResolutions = new List<TypeParameterResolutions>();
```

Comment:
Maybe `Expr` instead of `Exp` is a more common abbreviation of expression:

```suggestion
                internal IEnumerable<TypeParameterResolutions> ExprTypeParamResolutions = new List<TypeParameterResolutions>();
```

Reply:
done